quint/src/types/typeApplicationResolution.ts

```diff
@@ -0,0 +1,139 @@
+/* ----------------------------------------------------------------------------------
+ * Copyright 2024 Informal Systems
+ * Licensed under the Apache License, Version 2.0.
+ * See LICENSE in the project root for license information.
+ * --------------------------------------------------------------------------------- */
+
+/**
+ * Resolution of type application
+ *
+ * @author Shon Feder
+ *
+ * @module
+ */
+
+import { fail } from 'assert'
+import { FreshVarGenerator } from '../FreshVarGenerator'
+import { IRTransformer, transformType } from '../ir/IRTransformer'
+import { QuintTypeAlias } from '../ir/quintIr'
+import { QuintAppType, QuintType, QuintVarType, Row } from '../ir/quintTypes'
+import { LookupTable } from '../names/base'
+import { zip } from '../util'
+import { Substitutions, applySubstitution } from './substitutions'
+
+export class TypeApplicationResolver implements IRTransformer {
+  // Fresh variable generator, shared with the TypeInferrer
+  private freshVarGenerator: FreshVarGenerator
+  // Lookup table from the parser
+  private table: LookupTable
+
+  constructor(table: LookupTable, freshVarGenerator: FreshVarGenerator) {
+    this.table = table
+    this.freshVarGenerator = freshVarGenerator
+  }
+
+  exitType(t: QuintType): QuintType {
+    return this.resolveTypeApplications(t)
+  }
+
+  // Transforms `t` by resolving all the type applications in all its sub-terms
+  //
+  // E.g., given
+  //
+  //   type Foo[a, b] = (a, b)
+  //   type Bar[x, y] = {i: x, j: y}
+  //
+  //
+  // resolveTypeAllications(Foo[a, {f: Bar[int, str]}]) = (a, {f: {i: int, j: str}})
+  resolveTypeApplications(t: QuintType): QuintType {
+    const f: (_: QuintType) => QuintType = x => (x.kind !== 'app' ? x : this.resolveTypeApp(x))
+    return mapType(f, t)
+  }
+
+  private resolveTypeApp(t: QuintAppType): QuintType {
+    const typeDef = this.table.get(t.ctor.id!)! // TODO
```

Comment:
There is a TODO comment here with no description

Reply:
Yes. This branch is not quite ready for review yet!